magma/smart/smart_bits.py

```diff
@@ -0,0 +1,526 @@
+        if not isinstance(other, _SmartExpr):
+            return NotImplemented
+        return _SmartComparisonOpExpr(operator.lt, self, other)
+
+    # Shift operators.
+    def __lshift__(self, other: '_SmartExpr'):
+        if not isinstance(other, _SmartExpr):
+            return NotImplemented
+        return _SmartShiftOpExpr(operator.lshift, self, other)
+
+    def __rshift__(self, other: '_SmartExpr'):
+        if not isinstance(other, _SmartExpr):
+            return NotImplemented
+        return _SmartShiftOpExpr(operator.rshift, self, other)
+
+    # Unary operators.
+    def __invert__(self):
+        return _SmartUnaryOpExpr(operator.invert, self)
+
+    def __neg__(self):
+        return _SmartUnaryOpExpr(operator.neg, self)
+
+    # Reduction operators.
+    def reduce(self, op):
+        return _SmartReductionOpExpr(op, self)
+
+
+class _SmartOpExpr(_SmartExpr, metaclass=_SmartExprMeta):
+    def __init__(self, op, *args):
+        self._op = op
+        self._args = args
+
+    @property
+    def args(self):
+        return self._args
+
+    @property
+    def op(self):
+        return self._op
+
+    def _update(self, *args):
+        self._args = args
+
+    def _resolve_args(self, context):
+        for arg in self._args:
+            arg.resolve(context)
+
+    def _eval_args(self):
+        return [arg.eval() for arg in self._args]
+
+    def __str__(self):
+        if inspect.isbuiltin(self._op):
+            op_str = self._op.__name__
+        else:
+            op_str = str(self._op)
+        args = ", ".join(str(arg) for arg in self._args)
+        return f"{op_str}({args})"
+
+
+class _SmartExtendOpExpr(_SmartOpExpr):
+
+    class _ExtendOp:
+        def __init__(self, width):
+            self._width = width
+
+        def __call__(self, operand):
+            return operand.ext(self._width)
+
+        def __str__(self):
+            return f"Extend[{self._width}]"
+
+    def __init__(self, width, operand):
+        extend = _SmartExtendOpExpr._ExtendOp(width)
+        super().__init__(extend, operand)
+
+    def eval(self):
+        args = self._eval_args()
+        return self.op(*args)
+
+
+def _extend_if_needed(expr, to_width):
+    diff = expr._width_ - to_width
+    assert diff <= 0
+    if diff == 0:
+        return expr
+    signed = expr._signed_
+    expr = _SmartExtendOpExpr(-diff, expr)
+    expr._width_ = to_width
+    expr._signed_ = signed
+    return expr
+
+
+class _SmartReductionOpExpr(_SmartOpExpr):
+
+    class _ReductionOp:
+        _OP_TO_NAME = {
+            operator.and_: "And",
+            operator.or_: "Or",
+            operator.xor: "Xor",
+        }
+
+        def __init__(self, op):
+            cls = _SmartReductionOpExpr._ReductionOp
+            if op not in cls._OP_TO_NAME:
+                raise ValueError(f"Reduction operator {op} not supported")
+            self._op = op
+
+        def __call__(self, operand):
+            return reduce(self._op, operand)
```

Comment:
You may want to call `m.reduce` here since it will generate the coreir bitwise reduce primitive (which in turn generates something like `&x` in verilog)

Reply:
This is actually that function, there is `from magma.primitives.reduce import reduce` at the top of the file (maybe a `as magma_reduce` would be clearer).